File: app/src/main/java/info/horriblesubs/sher/adapter/ReleaseAdapter.java
package info.horriblesubs.sher.adapter;

import android.text.Html;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import info.horriblesubs.sher.R;
import info.horriblesubs.sher.api.horrible.model.ShowRelease;

public class ReleaseAdapter extends RecyclerView.Adapter<ReleaseAdapter.ViewHolder> {

    private final List<ShowRelease> listItems;
    private final OnItemClick onItemClick;
    private List<ShowRelease> items;
    private int size;

    public static ReleaseAdapter get(OnItemClick itemClick, List<ShowRelease> items, int s) {
        if (items == null) return new ReleaseAdapter(itemClick, null, 0);
        return new ReleaseAdapter(itemClick, items, Math.min(items.size(), s));
    }

    private ReleaseAdapter(OnItemClick onItemClick, @Nullable List<ShowRelease> items, int size) {
        if (items != null) this.listItems = new ArrayList<>(items);
        else this.listItems = new ArrayList<>();
        this.onItemClick = onItemClick;
        this.items = items;
        this.size = size;
    }

    public static ReleaseAdapter get(OnItemClick itemClick, List<ShowRelease> items) {
        return new ReleaseAdapter(itemClick, items, items == null?0:items.size());
    }

    @Override
    public int getItemViewType(int position) {
        return super.getItemViewType(position);
    }

    @NotNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.recycler_e, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, final int position) {
        holder.textView.setText(Html.fromHtml(items.get(position).release));
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onItemClick.onItemClicked(items.get(position));
            }
        });
        if (items.get(position).quality != null) {
            if (!items.get(position).quality.get(2)) holder.mark3.setVisibility(View.GONE);
            if (!items.get(position).quality.get(1)) holder.mark2.setVisibility(View.GONE);
            if (!items.get(position).quality.get(0)) holder.mark1.setVisibility(View.GONE);
        }
    }

    @Override
    public int getItemCount() {
        return size;
    }

    public interface OnItemClick {
        void onItemClicked(ShowRelease item);
    }

    public void onSearch(String s) {
        if (items == null) items = new ArrayList<>();
        if (s == null || s.isEmpty()) {
            items.clear();
            items.addAll(listItems);
            size = listItems.size();
            notifyDataSetChanged();
            return;
        }
        items.clear();
        for (ShowRelease item: listItems) {
            if (item.release.contains(s))
                items.add(item);
        }
        size = items.size();
        notifyDataSetChanged();
        Log.e("rls2.0", s + ", " + listItems.size());
        Log.e("rls2.0", s + ", " + size);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        private final AppCompatTextView textView;
        private final View mark1, mark2, mark3;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.textView);
            mark3 = itemView.findViewById(R.id.mark3);
            mark2 = itemView.findViewById(R.id.mark2);
            mark1 = itemView.findViewById(R.id.mark1);
        }
    }
}

File: app/src/main/java/info/horriblesubs/sher/ui/_extras/adapters/horrible/ReleaseAdapter.kt
package info.horriblesubs.sher.ui._extras.adapters.horrible

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.core.text.parseAsHtml
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.textview.MaterialTextView
import info.horriblesubs.sher.R
import info.horriblesubs.sher.data.subsplease.api.model.FHD
import info.horriblesubs.sher.data.subsplease.api.model.HD
import info.horriblesubs.sher.data.subsplease.api.model.ItemRelease
import info.horriblesubs.sher.data.subsplease.api.model.SD
import info.horriblesubs.sher.ui._extras.listeners.OnItemClickListener

class ReleaseAdapter(
    private val listener: OnItemClickListener<ItemRelease>
): RecyclerView.Adapter<ReleaseAdapter.Holder>(), Filterable {

    private var listClone = mutableListOf<ItemRelease>()
    private var list = mutableListOf<ItemRelease>()

    fun reset(list: List<ItemRelease>? = null) {
        this.list.clear()
        if (list.isNullOrEmpty()) {
            this.list = listClone
        } else {
            this.listClone.clear()
            this.list = list.toMutableList()
            this.listClone = list.toMutableList()
        }
        notifyDataSetChanged()
    }

    override fun getItemId(position: Int) = list[position].episode.toLongOrNull() ?: -1

    override fun getItemCount() = list.size

    override fun onCreateViewHolder(group: ViewGroup, viewType: Int) = Holder(
        LayoutInflater.from(group.context).inflate(R.layout.recycler_item_f, group, false),
        this@ReleaseAdapter
    )

    override fun onBindViewHolder(holder: Holder, position: Int) {
        holder.onBindToViewHolder(list[position], position)
    }

    @Suppress("UNCHECKED_CAST")
    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence?): FilterResults {
                val filteredList: List<ItemRelease> = if (charSequence.isNullOrBlank())
                    listClone
                else listClone.filter {
                    it.episode.contains(charSequence, true)
                }
                return FilterResults().apply {
                    values = filteredList
                }
            }

            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
                list.apply {
                    clear()
                    addAll(filterResults.values as? List<ItemRelease> ?: emptyList())
                }
                notifyDataSetChanged()
            }
        }
    }

    class Holder (
        view: View,
        private val adapter: ReleaseAdapter
    ): RecyclerView.ViewHolder(view) {

        private val textView: MaterialTextView = itemView.findViewById(R.id.textView)
        private val fhd: View = itemView.findViewById(R.id.mark3)
        private val hd: View = itemView.findViewById(R.id.mark2)
        private val sd: View = itemView.findViewById(R.id.mark1)

        fun onBindToViewHolder(t: ItemRelease, position: Int) {
            itemView.setOnClickListener{ adapter.listener.onItemClick(it, t, position) }
            if (t.FHD == null) fhd.visibility = View.GONE
            if (t.HD == null) hd.visibility = View.GONE
            if (t.SD == null) sd.visibility = View.GONE
            textView.text = t.episode.parseAsHtml()
        }
    }
}